File: app/src/main/java/com/thewhitewings/pouch/PouchApplication.java
package com.thewhitewings.pouch;

import android.app.Application;

import com.thewhitewings.pouch.data.DatabaseHelper;
import com.thewhitewings.pouch.data.NotesRepository;
import com.thewhitewings.pouch.data.OfflineNotesRepository;
import com.thewhitewings.pouch.data.PouchPreferences;
import com.thewhitewings.pouch.utils.Constants;

/**
 * The main application class for the Pouch app.
 * Custom app entry point for manual dependency injection.
 */
public class PouchApplication extends Application {

    // The repository of the app that manages the interactions with the databases and preferences.
    private NotesRepository notesRepository;

    @Override
    public void onCreate() {
        super.onCreate();

        // Initialize the database helper of the Creative zone
        DatabaseHelper creativeDatabaseHelper = new DatabaseHelper(
                this,
                Constants.CREATIVE_DATABASE_NAME,
                Constants.CREATIVE_AND_BOM_DATABASE_VERSION
        );

        // Initialize the database helper of the Box of mysteries zone
        DatabaseHelper bomDatabaseHelper = new DatabaseHelper(
                this,
                Constants.BOM_DATABASE_NAME,
                Constants.CREATIVE_AND_BOM_DATABASE_VERSION
        );

        // Initialize the SharedPreferences of the app
        PouchPreferences pouchPreferences = new PouchPreferences(this);

        // Initialize the notes repository with the database helpers and preferences
        notesRepository =
                new OfflineNotesRepository(
                        creativeDatabaseHelper,
                        bomDatabaseHelper,
                        pouchPreferences
                );
    }

    /**
     * Get the notes repository of the app.
     *
     * @return the notes repository of the app
     */
    public NotesRepository getNotesRepository() {
        return notesRepository;
    }
}

File: app/src/main/java/com/thewhitewings/pouch/PouchApplication.kt
package com.thewhitewings.pouch

import android.app.Application
import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.preferencesDataStore
import com.thewhitewings.pouch.feature_note.data.data_source.NoteDatabase
import com.thewhitewings.pouch.feature_note.data.repository.OfflineNotesRepositoryImpl
import com.thewhitewings.pouch.feature_note.data.repository.PouchPreferencesImpl
import com.thewhitewings.pouch.feature_note.data.util.Constants
import com.thewhitewings.pouch.feature_note.domain.repository.OfflineNotesRepository

// The DataStore instance of the app for storing preferences
private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(
    name = Constants.POUCH_PREFERENCE_NAME
)

/**
 * The main application class for the Pouch app.
 * Custom app entry point for manual dependency injection.
 */
class PouchApplication : Application() {

    /**
     * The repository of the app that manages the interactions with the databases and preferences.
     */
    lateinit var notesRepository: OfflineNotesRepository
        private set

    override fun onCreate() {
        super.onCreate()

        // Initialize the database of the Creative zone
        val creativeDatabase: NoteDatabase by lazy {
            NoteDatabase.getDatabase(this, Constants.CREATIVE_DATABASE_NAME)
        }

        // Initialize the database of the Box of mysteries zone
        val bomDatabase: NoteDatabase by lazy {
                NoteDatabase.getDatabase(this, Constants.BOM_DATABASE_NAME)
        }

        // Initialize the preference DataStore of the app
        val pouchPreferences = PouchPreferencesImpl(dataStore)

        // Initialize the notes repository with the databases and preferences
        notesRepository =
            OfflineNotesRepositoryImpl(
                creativeDatabase.noteDao(),
                bomDatabase.noteDao(),
                pouchPreferences
            )
    }
}
